### PR TITLE
MARATHON-8771 don't let to-be-restarted instances constrain placement

### PR DESCRIPTION
**Pull request by pierrebeitz**
Quoting timcharper:

> Currently, when Marathon restarts an app and evaluates constraints for
> placement of new instances, it only considers placed instances with the
> same configVersion. For example, if you have a hostname:UNIQUE
> constraint and perform a rolling upgrading, Marathon will allow two
> instances to be placed on the same node as long as their configVersion
> is different.
> When a Marathon app is restarted, however, the version is updated, but
> the configVersion remains the same. This means that Marathon will
> consider to-be-killed instances as part of the input for placements in
> constraints. For things like GROUP_BY, this is leading to the constraint
> ultimately being invalid.

This change seeks to change that behaviour by separating ScaleChanges
and RestartChanges and letting the latter modify `lastConfigChangeAt`.

JIRA issues:

* MARATHON-8771

### CODE REVIEW

**Comment by timcharper** [comment]
Hey there @pierrebeitz ! This is looking great so far !

Upon further consideration, I'm questioning whether or not it is really necessary to introduce a new version field. Originally, I thought we might do that in order to avoid breaking something or introducing confusion. But, upon further thought, I think that risk is minimal, and there is a lot of references in Marathon using ConfigRef which would need to be updated to RestartConfigRef.

In the end, I think we should just keep version and configVersion, and have restart update configVersion. This should be much simpler and probably less error prone.

**Comment by timcharper** [comment]
(I've updated the JIRA issue to reflect this direction)

**Review by timcharper — APPROVED** [review]

File: src/test/scala/mesosphere/marathon/state/VersionInfoTest.scala [hunk 1]
```diff
@@ -96,6 +113,28 @@ class VersionInfoTest extends UnitTest {
...
+
+      Then("lastConfigChangeAt is updated while lastScalingAt is not")
+      newVersion should be(
+        FullVersionInfo(
+          version = Timestamp(2),
+          lastScalingAt = Timestamp(1),
+          lastConfigChangeAt = Timestamp(2)
+        )
+      )
+      newVersion.lastConfigChangeVersion should equal(Timestamp(2))
```
Comment (pierrebeitz) [inline src/test/scala/mesosphere/marathon/state/VersionInfoTest.scala:135]: that's actually not needed once you realize that we only seem to use `VersionInfo` immutably. 

but better safe than sorry, i guess...

Comment (pierrebeitz) [inline src/test/scala/mesosphere/marathon/state/VersionInfoTest.scala:135]: explicitly checking `lastConfigChangeVersion` as that's the field `ResourceMatcher` uses to partition.

**Pull request merged**